### PR TITLE
boards/im880b: im880b board clock definition

### PR DESCRIPTION
**Pull request by jmfriedt**
Commit [c14d7ec7dbc2238ad251a70904407e503342b82d](https://github.com/RIOT-OS/RIOT/commit/c14d7ec7dbc2238ad251a70904407e503342b82d) requires the definition of the clock peripherals on a given board but the im880b does not provide such a definition in its configuration file. This PR defines the HSE clock as the 16 MHz external clock.
This PR aims at solving https://github.com/RIOT-OS/RIOT/pull/18293

Signed-off-by: Jean-Michel Friedt <friedtj@free.fr>

### CODE REVIEW

**Review by maribu — APPROVED** [review]
ACK. The code looks good and I trust your testing.

File: boards/im880b/include/periph_conf.h
Comment (maribu) [inline boards/im880b/include/periph_conf.h:33]:
```suggestion
/**
 * @brief   Speed of the HSE clock in Hz
 */
#define CLOCK_HSE                       MHZ(16)
```

File: boards/im880b/include/periph_conf.h
Comment (maribu) [inline boards/im880b/include/periph_conf.h:30]:
```suggestion
/**
 * @brief   This board has an HSE clock
 */
#define CONFIG_BOARD_HAS_HSE            1
```

**Review by maribu — COMMENTED** [review]
Oh, the CI wants the macros to be documented. (There is an exception for preexisting code that allowed us to update Doxygen and being more strict about non-documented stuff. That's why you may find other boards not having this documentation.)

Please squash the changes directly. You can use `git rebase -i <COMMIT_JUST_BEFORE_THE_FIRST_IN_PR>` and change the `pick` to `fixup` (or just `f` as shorthand).

Or even easier, commit the changes with `git commit --amend` to directly recreate the original commit, but including the changes you previously added. Finally, push with `git push --force-with-lease`.

(Note: You do not need to fear that the force-push closea the PR. This apparently happens when maintainers do so - which caused the trouble in the last PR - but the original author of a PR can do this without side effect.)

**Comment by maribu** [comment]
@jmfriedt ping ;)

I cannot merge this right now because the `static-tests` item in the CI tests fails due to the missing Doxygen documentation comment for `CONFIG_BOARD_HAS_HSE` and `CLOCK_HSE`. If you want, you can use the wording in my comments above or write your own comments instead. Afterwards commit with `git commit --amend` to amend the previous commit to also include the documentation rather than adding a second one. And, finally, push with `git push --force-with-lease`.

Sorry for the inconvenience.

**Comment by jmfriedt** [comment]
I am looking further in this issue because I think it is not as simple as I thought. I have two im880b boards, both programmed through the serial port using stm32-flash under GNU/Linux: one works like a charm, the other fails to start despite the clock configuration update. Both start fine when programmed using ST-Link/v2. Something is wrong, either the HSE clock is randomly initialized by the serial port programming and as a side effect of ST-Link programming is active, or the clock initialization sequence is wrong. RIOT-OS priori to September 2020 patch update would program both boards correctly and let them start HSE flawlessly, so I am checking if I am not missing some additional issue with the clock initialization which is all but easy to read (e.g. where is CLOCK_ENABLE_HSE defined? I understand include/periph_conf.h defines CONFIG_BOARD_HAS_LSE and CONFIG_BOARD_HAS_HSE but the clock initialization patch mentioned above uses CLOCK_ENABLE_HSE ... I cannot find the link between these constants).
Apologies for the delay, just trying to make sure I am unbreaking things correctly.

**Comment by jmfriedt** [comment]
I completed the operation after checking through all the patches involving the STM32L0L1 clock configuration that indeed defining HSE in the periph_conf.h is mandatory to get one board running but the fact that the second board remains stick when programming through the serial port is not satisfactory. As I was going through the patch history this second board ended working all the way to master before I unplugged and plugged back power which hints at a PLL configuration error that had remained from past tests until cold starting the board again. Someone knowledgeable with RIOT STM32 clock initialization should look into this, I cannot find the issue as I am not familiar enough (yet?) with the software architecture.

**Comment by maribu** [comment]
@aabadie is to my knowledge the person with the deepest insight into STM32 clock configuration, maybe he has an idea?

Given that the issue is only with one of two boards could indicate a hardware defect or a difference in revisions. Are you able to attach via SWD when one of the boards is stuck to pin down where exactly it is stuck? Would that board boot fine again after it is stuck due to flashing via serial/stm32flash when you do a cold boot? Maybe the issue is that the board doesn't get a reset after being flashed for whatever reason. (I have no idea why that would still work with the old release, though.)

Note: The comments seem to be still missing, so that the CI will still not let this pass.

**Comment by aabadie** [comment]
> where is CLOCK_ENABLE_HSE defined?

Here:
https://github.com/RIOT-OS/RIOT/blob/2e0746d11da74a81ff847d3b3329ab225e2e9b0e/cpu/stm32/stmclk/stmclk_l0l1.c#L260-L272

The idea is to only enable HSE if required by the configuration. `CLOCK_ENABLE_HSE` is only internal to the clock initialization implementation.
I have no idea what could be the issue either. I thought maybe some register was not reset correctly but apparently that's not the case:
https://github.com/RIOT-OS/RIOT/blob/2e0746d11da74a81ff847d3b3329ab225e2e9b0e/cpu/stm32/stmclk/stmclk_l0l1.c#L334-344
And this part of the code wasn't changed by c14d7ec7dbc2238ad251a70904407e503342b82d anyway.

**Comment by jmfriedt** [comment]
I have just received another 10 im880 boards from D. Donsez (Grenoble) so now I can make some statistics: of the 12 boards I have, 5 will NOT start executing the program when flashed using
```
stm32flash -w bin/im880b/program.bin /dev/ttyUSB0
```
et 7 will execute after programming and hitting the reset button. All boards will start executing the program when programmed using openocd/gdb (load and continue). After some trial and error, we (Gwen Goavec-Merou at FEMTO-ST) have concluded that adding at the end of ``stmclk_init_sysclk()`` of
``cpu/stm32/stmclk/stmclk_l0l1.c``
```
volatile int ;
for (i=0;i<2;i++) {__asm__("nop");}
```
sometimes solves the problem by allowing the program to execute. We have no idea why that is, but it matches the fact that refactoring the clock handling code led to the inability to execute code on some of the boards while pre-https://github.com/RIOT-OS/RIOT/commit/c14d7ec7dbc2238ad251a70904407e503342b82d was always working (and still is when using commit 6bf6b6be6c4723b49f62550a35111e57b7426aa4). Some asynchronous event seems to be leading to a race condition but we are unable to identify the cause beyond providing this "solution", or possibly some oscillator that is not settling fast enough when another peripheral requiring this clock source attempts at initializing. In the mean time I will use 6bf6b6be6c4723b49f62550a35111e57b7426aa4 for the coming field trip where I cannot afford have boards the sometimes start and sometimes fail, although I am quite unhappy at using such an outdated release. We have also verified that the content of the RCC registers (0x40023800 to 0x4002384c) are the same before and after the commit so as stated by aabadie, the functionality of the code seems not to have affected by refactoring the code.

**Comment by maribu** [comment]
Can you give a pointer on where to buy these boards? I only found one shop where they are out of stock.

Seems like an interesting challenge to debug the issue :)

**Comment by jmfriedt** [comment]
They must be https://www.soselectronic.com/products/imst/im880b-l-lorawan-234066 or https://shop.imst.de/wireless-modules/lora-products/77/im880b-l-prolink-lorawan-v3.0-long-range-radio?c=7 but I believe Didier Donsez at CS department of Grenoble University has some spare and since he sent me the ones I have, I can ship a couple for debugging.
I am not sure about the "interesting", but it is definitely challenging :)

**Comment by maribu** [comment]
Got two. Let's hope at least one is affected :)

**Pull request merged**
